buildSrc/src/main/java/org/elasticsearch/gradle/util/Util.java

```diff
@@ -75,4 +84,91 @@ public static URI getBuildSrcCodeSource() {
             throw new GradleException("Error determining build tools JAR location", e);
         }
     }
+
+    public static Function<FileTree, FileTree> filterByPatterns(
```

Comment:
This is a bit of an awkward way of defining a filter. Instead of creating a function that takes a `FileTree` and returns another `FileTree` why not model the filter as an `Action<? super PatternFilterable>`? Then your methods that take this could just look like:

```
getJavaMainSourceResources(project, spec -> spec.include(**/foo/**))
```

This also avoids the need to create a `PatternSet` via `PatternSetFactory` which is technically an internal API.

Reply:
Also would alleviate the need for a bunch of null-checking logic on the filter.

Reply:
Cool, thanks for the pointer. Updated to use `Action<? super PatternFilterable>` and think it is much cleaner now.